### ISSUE TITLE
Forward slash in JSON key treated as error

### ISSUE THREAD

**Issue by jakubklimek** [digitalbazaar/jsonld.js#543]
A forward slash in JSON key like this:
```json
{
  "@context": {
    "@version": 1.1,
    "a/b": "https://a"
  }
}
```
is treated as bad syntax: `jsonld.SyntaxError: Invalid JSON-LD syntax; term in form of IRI must expand to definition.`
However, forward slash is perfectly valid in JSON.
Where does this syntax error come from?

**Comment by jakubklimek** [comment]
However, Apache Jena riot also reports this as invalid:
```ERROR org.apache.jena.riot - A local context contains a term that has an invalid or missing IRI mapping [code=INVALID_IRI_MAPPING].```

**Comment by lemoustachiste** [comment]
@jakubklimek your issue is not too far from the one I was having here: https://github.com/digitalbazaar/jsonld.js/issues/542. Maybe it can help

**Comment by VladimirAlexiev** [comment]
@jakubklimek @lemoustachiste (and we got a similar issue at graphdb-support):

Forward slashes and colons are forbidden in JSON-LD terms (keys) because that will lead to confusion whether the term is an abbreviation or a URL.

Here's what the spec says https://w3c.github.io/json-ld-syntax/#terms :
> to avoid confusion between a [Compact IRI](https://w3c.github.io/json-ld-syntax/#dfn-compact-iri) and a [term](https://w3c.github.io/json-ld-syntax/#dfn-term), terms SHOULD NOT include a colon (:) and SHOULD be restricted to the form of [isegment-nz-nc](https://www.rfc-editor.org/rfc/rfc3987#section-2.2) as defined in [[RFC3987](https://w3c.github.io/json-ld-syntax/#bib-rfc3987)].

https://www.rfc-editor.org/rfc/rfc3987#section-2.2:
```
isegment-nz-nc = 1*( iunreserved / pct-encoded / sub-delims / "@" )
   ; non-zero-length segment without any colon ":"
iunreserved = ALPHA / DIGIT / "-" / "." / "_" / "~" / ucschar
ucschar = %xA0-D7FF / %xF900-FDCF / %xFDF0-FFEF / %x10000-1FFFD / %x20000-2FFFD / %x30000-3FFFD / %x40000-4FFFD /
  %x50000-5FFFD / %x60000-6FFFD / %x70000-7FFFD / %x80000-8FFFD / %x90000-9FFFD / %xA0000-AFFFD /
  %xB0000-BFFFD / %xC0000-CFFFD / %xD0000-DFFFD / %xE1000-EFFFD
sub-delims = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
```
As you see, forward slash is forbidden.

Let me illustrate the problem:
```json
{
  "@context": {
    "@version": 1.1,
    "@vocab": "https://example.org/",
    "a/b": "https://a"
  },
  "@id": "https://id",
  "a/b": 2 # is this https://a or https://example.org/a/b ?
}
```

What other punctuation is allowed in JSON-LD keys? You'd have to check all ranges of `ucschar`. 
I'm too lazy to do that, but the same RFC says
> Systems accepting IRIs MAY also deal with the printable characters in US-ASCII that are not allowed in URIs, namely "<", ">", '"', space, "{", "}", "|", "\", "^", and "`"

Because these punctuations are **forbidden** in URIs, I think that they are **allowed** in JSON-LD keys.

**Comment by jakubklimek** [comment]
@VladimirAlexiev fair point about the RFC3987 reference in the JSON-LD spec, I have missed that.

In your example, I would understand the problem with `"a:b"`, I do not understand it with `"a/b"` though, how could it be `https://example.org/a/b`? Only values with colons `:` are interpreted as compact IRIs, not with slashes.

**Comment by jakubklimek** [comment]
@VladimirAlexiev ahh, right. I forgot about `@vocab` and that in that case, JSON keys also directly represent parts of URIs.

Therefore, this is actually not an issue and can be closed.
